internal/v2/cert_verifier/cert_verifier_test.go

```diff
@@ -1,117 +1,221 @@
+			}
+			defer conn.Close()
+			c := s2av2pb.NewS2AServiceClient(conn)
+			log.Printf("Client: connected to: %s", address)
+			ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+			defer cancel()
+
+			// Setup bidrectional streaming session.
+			callOpts := []grpc.CallOption{}
+			cstream, err := c.SetUpSession(ctx, callOpts...)
+			if err != nil  {
+				log.Fatalf("Client: failed to setup bidirectional streaming RPC session: %v", err)
+			}
+			log.Printf("Client: set up bidirectional streaming RPC session.")
+
+			VerifyPeerCertificateFunc := VerifyClientCertificateChain(cstream)
+			got, want := VerifyPeerCertificateFunc(tc.rawCerts, nil), tc.expectedErr
+			if want == nil {
+				if got != nil {
+					t.Errorf("Peer Certificate verification failed, got: %v, want: %v", got, want)
+				}
+			} else {
+				if got == nil {
+					t.Errorf("Peer Certificate verification failed, got: %v, want: %v", got, want)
+				}
+				if got.Error() != want.Error() {
+					t.Errorf("Peer Certificate verification failed, got: %v, want: %v", got, want)
+				}
 			}
 		})
 	}
+	stop()
 }
 
 // TestVerifyServerCertChain runs unit tests for VerifyServerCertificateChain.
 func TestVerifyServerCertChain(t *testing.T) {
-	// Setup for static server test : empty certPool.
-	certPool0 := x509.NewCertPool()
-
-	// Setup for static server test : non-empty certPool.
-	certPool1 := x509.NewCertPool()
-	certPool1.AppendCertsFromPEM(clientCert)
-	cert1, err1 := tls.X509KeyPair(clientCert, clientKey)
-	if err1 != nil {
-		t.Errorf("Test suite setup failed: %v", err1)
+	// Start up fake S2Av2 server.
+	var wg sync.WaitGroup
+	wg.Add(1)
+	stop, address, err := startFakeS2Av2Server(&wg)
+	wg.Wait()
+	if err != nil {
+		log.Fatalf("error starting fake S2Av2 Server: %v", err)
 	}
 
 	for _, tc := range []struct {
 		description              string
-		pool                     *x509.CertPool
-		expectedCommonName       string
 		hostname                 string
 		rawCerts                 [][]byte
-		verifiedChains           [][]*x509.Certificate
-		err                      error
+		expectedErr              error
 	}{
 		{
-			description: "static : empty certPool",
-			pool: certPool0,
-			expectedCommonName: "s2a_test_cert",
+			description: "empty chain",
 			hostname: "host",
 			rawCerts: nil,
-			verifiedChains: nil,
-			err: nil,
+			expectedErr: errors.New("rpc error: code = Unknown desc = Server Peer Verification failed: server cert chain is empty."),
 		},
 		{
-			description: "static : non-empty certPool",
-			pool: certPool1,
-			expectedCommonName: "s2a_test_cert",
+			description: "chain of length 1",
 			hostname: "host",
-			rawCerts: cert1.Certificate,
-			verifiedChains: [][]*x509.Certificate{{cert1.Leaf},{}},
-			err: nil,
+			rawCerts: [][]byte{serverRootDERCert, },
+			expectedErr: nil,
+		},
+		{
+			description: "chain of length 2_correct",
+			hostname: "host",
+			rawCerts: [][]byte{serverIntermediateDERCert, serverRootDERCert, },
```

Comment:
Can we use `{serverLeafDERCert, serverIntermediateDERCert}` here?

Reply:
After tracing through this again, rawCerts is passed to CertificateChain in the proto: https://github.com/google/s2a-go/blob/23127301b4c006bbb890458e74ce23f8d64cf719/internal/proto/v2/s2a_go_proto/s2a.pb.go#L1524

The CertificateChain field has to include certs all the way from leaf to root. I changed this to a chain of 3: leaf->intermediate->root in both the Client Peer test and Server Peer test.

To align with the above fact, I also added to the fake rpc implementation:
Since we are expecting the CertificateChain in the req to contain leaf to root inclusive, I now check that the root given to us in the certificate chain(last element in cert chain) matches one of the roots in S2A(hardcoded since this is a fake)

Reply:
Discussed offline:
- Correct documentation in proto file to make clear that CertificateChain does NOT include root
- Remove the extra check added to the fake RPC